abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1969 @@
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Calls registered foreign function (`name_data`, `name_size`)
+with arguments (`arguments_data`, `arguments_size`).
+
+Return value(s) (`return_results_data`, `return_results_size`)
+are optional.
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when the requested function was not found.
+- `INVALID_MEMORY_ACCESS` when `name_data`, `name_size`,
+  `arguments_data`, `arguments_size`, `return_results_data`
+  and/or `return_results_size` point to invalid memory address.
+
+
+### Callbacks exposed by the Wasm module
+
+#### `proxy_on_foreign_function`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (uint32_t) function_id`
+  - `i32 (size_t) arguments_size`
+* returns:
+  - none
+
+Called when a registered foreign callback `function_id` is called.
+
+Its arguments (of `arguments_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to
+`FOREIGN_FUNCTION_ARGUMENTS`.
+
+
+## Unimplemented WASI functions
+
+Various unimplemented WASI functions that are expected to be present
+when the Wasm module is compiled for the `wasm32-wasi` target.
+
+
+### Functions exposed by the host
+
+#### `wasi_snapshot_preview1.args_sizes_get`
+
+* params:
+  - `i32 (size_t *) return_argc`
+  - `i32 (size_t *) return_argv_buffer_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Hosts must write `0` to both output arguments to prevent
+[`wasi_snapshot_preview1.args_get`] from being called.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `FAULT` when `return_argc` and/or `return_argv_buffer_size` point to
+  invalid memory address.
+
+
+#### `wasi_snapshot_preview1.args_get`
+
+* params:
+  - `i32 (uint8_t **) return_argv`
+  - `i32 (uint8_t *) return_argv_size`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+This function should not be called if zero sizes were returned
+in [`wasi_snapshot_preview1.args_sizes_get`].
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+
+
+#### `wasi_snapshot_preview1.proc_exit`
+
+* params:
+  - `i32 (uint32_t) exit_code`
+* returns:
+  - none
+
+This function is never called.
+
+
+# Types
+
+#### `proxy_log_level_t`
+
+- `TRACE` = `0`
+- `DEBUG` = `1`
+- `INFO` = `2`
+- `WARN` = `3`
+- `ERROR` = `4`
+- `CRITICAL` = `5`
+
+
+#### `proxy_status_t`
+
+- `OK` = `0`
+- `NOT_FOUND` = `1`
+- `BAD_ARGUMENT` = `2`
+- `SERIALIZATION_FAILURE` = `3`
+- `PARSE_FAILURE` = `4`
+- `INVALID_MEMORY_ACCESS` = `6`
+- `EMPTY` = `7`
+- `CAS_MISMATCH` = `8`
+- `INTERNAL_FAILURE` = `10`
+- `UNIMPLEMENTED` = `12`
+
+
+#### `proxy_action_t`
+
+- `CONTINUE` = `0`
+- `PAUSE` = `1`
+
+
+#### `proxy_buffer_type_t`
+
+- `HTTP_REQUEST_BODY` = `0`
+- `HTTP_RESPONSE_BODY` = `1`
+- `DOWNSTREAM_DATA` = `2`
+- `UPSTREAM_DATA` = `3`
+- `HTTP_CALL_RESPONSE_BODY` = `4`
+- `GRPC_CALL_MESSAGE` = `5`
+- `VM_CONFIGURATION` = `6`
+- `PLUGIN_CONFIGURATION` = `7`
+- `FOREIGN_FUNCTION_ARGUMENTS` = `8`
+
+
+#### `proxy_header_map_type_t`
+
+- `HTTP_REQUEST_HEADERS` = `0`
```

Comment:
Hey @PiotrSikora! Quick question: has it been considered to make URI arguments a map type as well? How are proxy-wasm users currently expected to manipulate that part of the request? Through platform.implementation-specific properties maybe? Would it make sense to introduce a  `HTTP_REQUEST_URI_ARGS` type?